1_IGT_PP/Code/R/1_Analyses/correlations_testretest_joint.Rmd

```diff
@@ -0,0 +1,177 @@
+---
+title: "Joint Model Test-Retest Plots"
+output: html_document
+date: "2023-03-15"
+---
+
+
+
+# ---------------------------------------
+# Setup
+## Loading Packages & Data
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+
+library(tidyverse)
+library(here)
+library(lemon)
+library(rstan)
+library(hBayesDM)
+library(ggh4x)
+library(hypergeo)
+source(here("1_IGT_PP", "Code", "R", "1_Analyses", "plausible_correlations.R"))
+
+blups = readRDS(here("1_IGT_PP", "Data", "2_Fitted", "Joint", "orl_pp_joint_blups.rds"))
+
+correlations = readRDS(here("1_IGT_PP", "Data", "2_Fitted", "Joint", "orl_pp_joint_model_correlations.rds"))
+
+orl_fit = readRDS(here("1_IGT_PP", "Data", "2_Fitted", "Joint", "orl_pp_joint_fit.rds"))
+```
+
+
+
+# ---------------------------------------
+# Population Corrections
+```{r}
+# Extract posterior samples for each parameter
+posterior = rstan::extract(orl_fit)
+```
+
+
+
+```{r}
+# Create empty dataframe
+iterative_correlations = data.frame()
+
+
+for(param in c("Arew", "Apun", "K", "betaF", "betaP")){
+  for(i in 1:nrow(posterior[[param]])){
+    # Correlate parameter value between session 1 and session 2 on iteration i
+    current_corr = data.frame(parameter = param,
+                              correlation = cor(posterior[[param]][i, , 1],
+                                                posterior[[param]][i, , 2]))
+    
+    # Bind this iteration's correlation with all the other correlations
+    iterative_correlations = bind_rows(iterative_correlations,
+                                       current_corr)
+  }
+}
+```
+
+
+
+```{r}
+# Create empty dataframe
+corrected_iterative_corrs = data.frame()
+
+for(param in c(#"Arew", # doesn't work for some reason
+               "Apun",
+               # "K",   # doesn't work for some reason
+               "betaF", "betaP")){
+  # subset correlations for parameter
+  current_param = filter(iterative_correlations, parameter == param)
+  
+  # apply population correction
+  current_corrected = data.frame(parameter = param,
+                                 correlation = pop_correct_MCMC(current_param$correlation,
+                                                                n_subj = 49, kappa = 1))
```

Comment:
OK so I figured out what was wrong here—this correction relies on a few different methods that are quite sensitive to when the original correlation estimate (or, the given iteration of the estimate given we have a bunch of MCMC samples of the correlations) is close to the parameter bounds of -1 or 1. We ran into this before, but correlations were close to 1, not -1. As such, my fix did not work for `K` and `Arew` here because there are some values that are right up at -1. I have come up with a new fix that does the trick by rescaling the input correlations so that they are not up against bounds, and then scaling results back at the end. It does produce a small amount of approximation error, but I am not sure of another solution. I will leave a suggestion below that you can commit to get this code working. 

One other note here—we do not actually need to apply the correction to `betaF` and `betaP`. The correction is only necessary when the correlation we are estimating is on the parameters when they have a different scale than the scale we interpret them on. e.g., the correlation on `Arew` and `Apun` is estimated on the scale before `Phi_approx(...)`—before they are transformed to be between 0 and 1. Therefore, if we want the correlation on the learning parameters when on the 0-1 scale, we must do the iterative correlation step and then the population correction to obtain the implied correlation on the scale we care about. Conversely, `betaF` and `betaP` are not transformed, so correlation straight from the model (e.g., `posterior[["betaF"]][,1,2]`) is already the quantity we want. 

In theory, this means that the `pop_correct_MCMC` method applied to `betaF` and `betaP` should give essentially the same result (with some approximation error) as the actual correlation extracted directly from the model. I will include some code in the comment I leave below to demonstrate this!